### PR TITLE
`fn angular_ipred`: Make safe w/ `Rav1dPictureDataComponentOffset`

### PR DESCRIPTION
**Pull request by kkysen**
This makes `fn angular_ipred` and all of its fallbacks fully safe.

For most of these, this was pretty simple, but `fn ipred_filter_rust` was more complex since it uses normal buffers for the first iteration and then switches to `dst: Rav1dPictureDataComponentOffset`.  For the inner of these, I switched to an `if`, which as I explain in the commit 14dde20, I think should be fine perf-wise.

As for the call sites, sometimes and tmp/scratch buffer is passed in instead of an existing `Rav1dPictureDataComponent`, so I added `fn wrap_buf` for this, which safely wraps a `&mut buf`.  Since this is a unique reference, subdividing it further with the `DisjointMut` is sound (normally the buf is owned by the `DisjointMut`, and so is unique that way).

### CODE REVIEW

File: include/dav1d/picture.rs [hunk 1]
```diff
@@ -207,6 +221,12 @@ unsafe impl AsMutPtr for Rav1dPictureDataComponentInner {
 pub struct Rav1dPictureDataComponent(DisjointMut<Rav1dPictureDataComponentInner>);
 
 impl Rav1dPictureDataComponent {
+    pub fn wrap_buf<BD: BitDepth>(buf: &mut [BD::Pixel], stride: usize) -> Self {
```
Comment (randomPoison) [inline include/dav1d/picture.rs:224]: We're not tying the lifetime of the new `Rav1dPictureDataComponent{Inner}` to the lifetime of the buffer, right? Could we add a lifetime to `Rav1dPictureDataComponent` so that we can ensure that it doesn't outlive the borrowed buffer? The lifetime could be `'static` for owned picture data components. I suspect that'd make the type much more cumbersome to work with though, since we'd need to add lifetime annotations pretty much everywhere this is used.

If that won't work, should we at least add a safety comment here saying that the `Rav1dPictureDataComponent` can't outlive `buf`? I feel like that should make this function `unsafe` as well, since it creates a borrow that's not tracked by the compiler.

File: src/ipred.rs [hunk 2]
```diff
@@ -2206,6 +2282,9 @@ impl Rav1dIntraPredDSPContext {
     }
 
     pub const fn new<BD: BitDepth>(flags: CpuFlags) -> Self {
-        Self::default::<BD>().init::<BD>(flags)
+        let default = Self::default::<BD>();
+        let mut c = Self::default::<BD>().init::<BD>(flags);
+        c.intra_pred[FILTER_PRED as usize] = default.intra_pred[FILTER_PRED as usize];
```
Comment (randomPoison) [inline src/ipred.rs:2287]: What's this change for? It seems like it's unrelated to the rest of the changes in the PR, and it's unclear why this is even necessary. This seems like it's overriding the value of `c.intra_pred[FILTER_PRED]` set in `init` with the default value, i.e. the Rust fallback. Is that the correct thing to do? Isn't that going to override the asm routine that was set in `init`?

File: include/dav1d/picture.rs (hunk 1 above)
Comment (kkysen) [reply include/dav1d/picture.rs:224]: Yeah, the lifetime is incorrect/unchecked right now.  @rinon brought that up, too.  I think we can add a lifetime that defaults to `'static`.  Can I add that as a PR on top of everything, though, since there's a lot on top of this?

File: src/ipred.rs (hunk 2 above)
Comment (kkysen) [reply src/ipred.rs:2287]: Whoops, I added this for debugging/checking I didn't break anything while running tests faster.  I did not mean to commit this.

File: src/ipred.rs [hunk 3]
```diff
@@ -1200,7 +1200,8 @@ unsafe fn ipred_filter_rust<BD: BitDepth>(
     for y in (0..height).step_by(2) {
         let topleft_off = topleft_off - y;
         let mut topleft = topleft_in[topleft_off];
-        for x in (0..width).step_by(4) {
+        for x in 0..width / 4 {
+            let x = x * 4;
```
[the remark this answers is not in the document]
Comment (kkysen) [reply src/ipred.rs:1272]: Hmm, I might be totally wrong.  I forgot exactly what I was trying to optimize.

[the remark this answers is not in the document]
Comment (kkysen) [reply src/ipred.rs:1272]: Oh yeah, I think I wanted to make sure `width` is divisible by 4, so that `top[x..][..4]` where `top.len() == width` doesn't need a check.  If `width % 4 != 0`, then `x` could be `width - 1`, but if `width % 4 == 0`, then it guarantees that the max `x` is `width - 4`.  Let me fix this.

File: include/dav1d/picture.rs (hunk 1 above)
Comment (randomPoison) [reply include/dav1d/picture.rs:224]: Yeah, I'm fine doing that in a separate PR 👍

File: src/ipred.rs (hunk 2 above)
Comment (kkysen) [reply src/ipred.rs]: Fixed in [`c8c8921..cdc3c5e`](https://github.com/memorysafety/rav1d/compare/c8c89218a1ffb4537d12bf56e463d5f04c5b039d..cdc3c5e32b4e84bd0da00dca204bea2b740d1a60).

File: src/ipred.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kkysen) [reply src/ipred.rs]: Fixed in [`cdc3c5e..59ab627`](https://github.com/memorysafety/rav1d/compare/cdc3c5e32b4e84bd0da00dca204bea2b740d1a60..59ab62729478ac84d58b19b8460ed41612dc3694).

**Review by randomPoison — APPROVED** [review]
